controllers/mirroring/mirroring_controller.go

```diff
@@ -0,0 +1,478 @@
+				}
+				return nil
+			})
+			if err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to create/update bootstrap secret: %w", err)
+			}
+		}
+
+		_, err = controllerutil.CreateOrUpdate(r.ctx, r.Client, cephBlockPool, func() error {
+			util.AddAnnotation(
+				cephBlockPool,
+				util.BlockPoolMirroringTargetIDAnnotation,
+				response.BlockPoolsInfo[i].BlockPoolID,
+			)
+
+			if disableMirroring {
+				cephBlockPool.Spec.Mirroring = rookCephv1.MirroringSpec{}
+			} else {
+				cephBlockPool.Spec.Mirroring.Enabled = true
+				cephBlockPool.Spec.Mirroring.Mode = "image"
+				if mirroringToken != "" {
+					if cephBlockPool.Spec.Mirroring.Peers == nil {
+						cephBlockPool.Spec.Mirroring.Peers = &rookCephv1.MirroringPeerSpec{SecretNames: []string{}}
+					}
+
+					if !slices.Contains(cephBlockPool.Spec.Mirroring.Peers.SecretNames, secretName) {
+						cephBlockPool.Spec.Mirroring.Peers.SecretNames = append(
+							cephBlockPool.Spec.Mirroring.Peers.SecretNames, secretName)
+					}
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf(
+				"failed to update CephBlockPool %s for mirroring: %w",
+				cephBlockPool.Name,
+				err,
+			)
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *MirroringReconciler) reconcileRadosNamespaceMirroring(
+	ocsClient *providerClient.OCSProviderClient,
+	clientMappingConfig *corev1.ConfigMap,
+	storageClusterPeer *ocsv1.StorageClusterPeer,
+	disableMirroring bool,
+) (ctrl.Result, error) {
+	/*
+		Algorithm:
+		make a list of peerClientIDs
+		send GetStorageClientsInfo with this Info
+		make a map of remoteNamespaceByClientID from response
+		list all radosNamespace and for each,
+			find out which consumer does it belong to
+			find the localClientID and use map to get radosnamespce
+			if radosnamespace is empty, disable mirroring
+			else enable mirroring
+	*/
+
+	peerClientIDs := []string{}
+	storageConsumerByName := map[string]*ocsv1alpha1.StorageConsumer{}
+	for localClientID, peerClientID := range clientMappingConfig.Data {
+		// for internal mode, we need only blockPool mirroring, hence skipping this for the special key "internal"
+		if localClientID == internalKey {
+			continue
+		}
+		// Check if the storageConsumer with the ClientID exists, this is a fancy get operation as
+		// there will be only one consumer with the clientID
+		storageConsumers := &ocsv1alpha1.StorageConsumerList{}
+		if err := r.list(storageConsumers, client.MatchingFields{clientIDIndexName: localClientID}); err != nil {
+			return ctrl.Result{}, err
+		}
+		if len(storageConsumers.Items) != 1 {
+			return ctrl.Result{}, fmt.Errorf("expected 1 StorageConsumer but got %v", len(storageConsumers.Items))
+		}
+		storageConsumerByName[storageConsumers.Items[0].Name] = &storageConsumers.Items[0]
+		peerClientIDs = append(peerClientIDs, peerClientID)
+	}
+
+	response, err := ocsClient.GetStorageClientsInfo(
+		r.ctx,
+		storageClusterPeer.Status.PeerInfo.StorageClusterUid,
```

Comment:
Shouldn't we first check if we have a peerInfo with a storageClusterUid value before we send a request. Maybe we even need this to this check as the first check in the function